sled-agent/src/bootstrap/agent.rs

```diff
@@ -486,16 +652,26 @@ impl Agent {
     ) -> Result<SledAgentResponse, BootstrapError> {
         info!(&self.log, "Loading Sled Agent: {:?}", request);
 
+        // Initialize the secret retriever used by the `KeyManager`
+        if self.bootstore.get_status().await?.fsm_state == "uninitialized" {
+            // We aren't using LRTQ as this a cluster with fewer than 3 sleds
```

Comment:
This kinda seems like a leap-of-logic here -- it might be right, but it isn't obvious to me.

Why would the bootstore being "uninitialized" imply this?

Reply:
Another good find. I should definitely add a comment here.

We only call `RequestSledAgent` when a `SledAgentRequest` has been noticed by the bootstrap agent either directly received from RSS or found in the ledger on cold boot. RSS always calls agent`bootstore::init_rack` before any `SledAgentRequest` gets sent and must successfully complete for the `SledAgentRequest`s to be sent. `bootstore::rack_init` can only complete if every single sled has moved to `Initial_member` state. 

Therefore we can be sure that if we are still in the `uninitialized` state when `request_sled_agent` has been called that we had fewer than 3 sleds and the trust quorum wasn't used because we must have skipped over `bootstore::rack_init`.

See https://github.com/oxidecomputer/omicron/blob/09c6faab7770be71d62d5dd91ccd75a36d87b6a5/sled-agent/src/rack_setup/service.rs#L882-L902 for the ordering.

Now, again, the use of this method on rack_reset has me concerned that 'de-initialization' doesn't occur. I think that actually may not be a problem as long as we do a reboot after rss reset, but it still seems racy to me, as U.2. disks can be upserted right after the reset.

Reply:
I agree that "de-initialization" seems like it should be a thing? Maybe we should wipe the bootstore in rack reset?

But also, that still seems like a way of justifying "this is what the code happens to do today". Isn't the decision to be "hard-coded" or not very known to the bootstore? Like, don't we have an `is_hardcoded` method somewhere? Why not use **that** to make this call?

Reply:
> Maybe we should wipe the bootstore in rack reset?

We *have* to wipe the bootstore on reset. That's the only way it's safe. We either succeed on rack_init and every single initial member is initialized or we have to wipe it. I had assumed prior to this that we were doing a *factory* reset, which would clear the M.2s, as we discussed in the update sync meetings a few months ago. But it looks like we are leaving the sled agent running after we do the wipe, which is a cause for concern to me. If we wiped the M.2s and restarted the process state would then reload from the M.2s and the bootstore would go back to being uninitialized. I'm concerned that we are trying to unwind a bunch of mutable state not just in the bootstore but all throughout the sled agent, and there's no real good way to tell that this is correct.

> But also, that still seems like a way of justifying "this is what the code happens to do today". Isn't the decision to be "hard-coded" or not very known to the bootstore? Like, don't we have an is_hardcoded method somewhere? Why not use that to make this call?

The bootstore doesn't know anything about the key manager or the sled-agent, or disk encryption keys. All it knows about are key shares and how to go about reconstructing a rack secret with those shares. It provides that rack secret to the sled agent which then is able to derive various keys from it using HKDF and other algorithms. Right now the only keys are disk encryption keys, but we also plan to derive the root private keys for our internal CA from the rack secret in the future. 

The `HardcodedSecretRetriever` is just another implementation of the `SecretRetriever` trait that is used to provide input key material (IKM) to the `KeyManager` so that it can derive encryption keys. The bootstore does not know anything about it. This is also why we've been able to be running with ZFS encryption turned on this whole time and none of the code I touched changes how that works. It just changes the raw secret input to the whole thing. 

Now, since we effectively know how to build an arbitrary key-value store on top of the bootstore, we *could* put extra information in a key related to this, but that wouldn't be known to lrtq, it would just be arbitrary data that the sled-agent could read and write. Maybe that's the right decision. But then again, we probably don't even need to do that since we can put that information in the persisted `SledAgentRequest` itself. I can add that information there in this PR, and I think that would be a good idea to clean up this somewhat convoluted logic leap, that only makes sense with a deep understanding of how the bootstore and sled-agent are ordered together.

Reply:
I added a boolean to `SledAgentRequest` and used that to decide whether to use trust quorum in 1607e8bdaf00a8e22832a77b83bf5cdf10afdb59

The code is *much* more clear now. 

However, I still remain concerned about trying to reset the memory state of the process. I really think we should reconsider that. An RSS reset command should, IMO, wipe the M.2s and take the reboot the sled. Practically speaking this is really what we do, since a mupdate immediately takes the sleds down to A2 anyway before it does anything.